app/v2/controllers/V2MovementsController.scala

```diff
@@ -0,0 +1,328 @@
+          .fold(
+            presentationError => Status(presentationError.code.statusCode)(Json.toJson(presentationError)),
+            response => Ok(Json.toJson(HateoasMovementIdsResponse(response, movementType, updatedSince)))
+          )
+    }
+
+  def attachMessage(movementType: MovementType, movementId: MovementId): Action[Source[ByteString, _]] =
+    contentTypeRoute {
+      case Some(MimeTypes.XML)  => attachMessageXML(movementId, movementType)
+      case Some(MimeTypes.JSON) => attachMessageJSON(movementId, movementType)
+    }
+
+  private def attachMessageXML(movementId: MovementId, movementType: MovementType): Action[Source[ByteString, _]] =
+    (authActionNewEnrolmentOnly andThen messageSizeAction()).streamWithAwait {
+      awaitFileWrite => implicit request =>
+        implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
+
+        val messageTypeList =
+          if (movementType == MovementType.Arrival) MessageType.updateMessageTypesSentByArrivalTrader else MessageType.updateMessageTypesSentByDepartureTrader
+
+        (for {
+          messageType <- xmlParsingService.extractMessageType(request.body, messageTypeList).asPresentation
+          _           <- awaitFileWrite
+          _           <- validationService.validateXml(messageType, request.body).asPresentation
+          _ = auditService.audit(messageType.auditType, request.body, MimeTypes.XML)
+          updateMovementResponse <- movementsService.updateMovement(movementId, movementType, messageType, request.body).asPresentation
+          _ = pushNotificationsService.update(movementId)
+          _ <- routerService
+            .send(messageType, request.eoriNumber, movementId, updateMovementResponse.messageId, request.body)
+            .asPresentation
+        } yield updateMovementResponse).fold[Result](
+          presentationError => Status(presentationError.code.statusCode)(Json.toJson(presentationError)),
+          response => Accepted(Json.toJson(HateoasMovementUpdateResponse(movementId, response.messageId, movementType)))
+        )
+    }
+
+  private def attachMessageJSON(id: MovementId, movementType: MovementType): Action[Source[ByteString, _]] = {
+
+    def handleXml(movementId: MovementId, eoriNumber: EORINumber, messageType: MessageType, src: Source[ByteString, _])(implicit
+      hc: HeaderCarrier
+    ): EitherT[Future, PresentationError, UpdateMovementResponse] =
+      withReusableSource(src) {
+        source =>
+          for {
+            _              <- validationService.validateXml(messageType, source).asPresentation(jsonToXmlValidationErrorConverter, materializerExecutionContext)
+            updateResponse <- movementsService.updateMovement(movementId, movementType, messageType, source).asPresentation
+            _ = pushNotificationsService.update(movementId)
+            _ <- routerService.send(messageType, eoriNumber, movementId, updateResponse.messageId, source).asPresentation
```

Comment:
I made a change in a previous PR to ensure that this logic (which is the same for XML and Json routes) was put in its own method -- it seems to have been removed here and we have duplication across both routes -- can we put the single method back?